catroid/src/main/java/org/catrobat/catroid/stage/TextToSpeechHolder.kt

```diff
@@ -0,0 +1,103 @@
+import org.catrobat.catroid.content.bricks.Brick
+import java.io.File
+import java.util.HashMap
+
+class TextToSpeechHolder private constructor() {
+
+    fun initTextToSpeech(stageActivity: StageActivity, stageResourceHolder: StageResourceHolder) {
+        textToSpeech = TextToSpeech(stageActivity, OnInitListener { status ->
+                if (status == TextToSpeech.SUCCESS) {
+                    utteranceProgressListenerContainer = UtteranceProgressListenerContainer()
+                    textToSpeech?.setOnUtteranceProgressListener(utteranceProgressListenerContainer)
+                    stageResourceHolder.resourceInitialized()
+                } else {
+                    val builder = AlertDialog.Builder(ContextThemeWrapper(stageActivity, R.style.Theme_AppCompat_Dialog))
+                    builder.setMessage(R.string.prestage_text_to_speech_engine_not_installed)
+                        .setCancelable(false)
+                        .setPositiveButton(R.string.yes) { dialog, id ->
+                            val installIntent = Intent()
+                            installIntent.action = TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA
+                            stageActivity.startActivity(installIntent)
+                            stageResourceHolder.resourceFailed(Brick.TEXT_TO_SPEECH)
+                        }
+                        .setNegativeButton(R.string.no) { dialog, id ->
+                            stageResourceHolder.resourceFailed(Brick.TEXT_TO_SPEECH)
+                        }
+                    val alert = builder.create()
+                    alert.show()
+                }
+            })
+    }
+
+    fun shutDownTextToSpeech() {
+        if (textToSpeech != null) {
+            textToSpeech?.stop()
+            textToSpeech?.shutdown()
+        }
+    }
+
+    fun textToSpeech(text: String?, speechFile: File, listener: UtteranceProgressListener, speakParameter: HashMap<String?, String?>) {
+        val text = text ?: ""
+        if (utteranceProgressListenerContainer!!.addUtteranceProgressListener(speechFile,
+            listener, speakParameter[TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID]!!)) {
+            val status = textToSpeech?.synthesizeToFile(text, speakParameter, speechFile.absolutePath)
+            if (status == TextToSpeech.ERROR) {
+                Log.e(TAG, "File synthesizing failed")
+            }
+        }
+    }
+
+    fun deleteSpeechFiles() {
+        val pathToSpeechFiles = File(Constants.TEXT_TO_SPEECH_TMP_PATH)
+        if (pathToSpeechFiles.isDirectory) {
+            pathToSpeechFiles.walk().forEach {
+                it.delete()
+            }
+        }
+    }
+
+    companion object {
+        private val TAG = TextToSpeechHolder::class.java.simpleName
+        private var textToSpeech: TextToSpeech? = null
+        private var utteranceProgressListenerContainer: UtteranceProgressListenerContainer? = null
+        @JvmStatic
+        var instance: TextToSpeechHolder? = null
```

Comment:
Please use "val" instead
val instance = TextToSpeechHolder() 
(much less code and not a nullable type)

Reply:
Since this is a Singelton you can use "object" as well.

https://kotlinlang.org/docs/reference/object-declarations.html#object-declarations